File: src/lib/norgesbank.ts
// Norges Bank API — corrected dataset and key format
// GOVT_GENERIC_RATES: business daily yields
//   TBIL (Treasury Bills): 3M, 6M, 12M
//   GBON (Government Bonds): 3Y, 5Y, 7Y, 10Y
// GOVT_ZEROCOUPON: zero-coupon yields (includes 2Y, useful for steepness/spreads)

const NB_BASE = "https://data.norges-bank.no/api/data";

export interface NBPoint {
  date: string;
  value: number;
}

// Generic SDMX-JSON parser for Norges Bank responses
async function fetchSDMX(url: string): Promise<NBPoint[]> {
  const res = await fetch(url, { next: { revalidate: 0 } });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`Norges Bank request failed ${res.status}: ${url}\n${body.slice(0, 200)}`);
  }
  const json = await res.json();
  const struct = json.data.structure;
  const ds = json.data.dataSets[0];
  const series = ds.series as Record<string, { observations: Record<string, number[]> }>;

  // Find the time dimension in the observation dimensions
  const obsDims: { id: string; values: { id: string }[] }[] = struct.dimensions.observation;
  const timeDim = obsDims.find((d) => d.id === "TIME_PERIOD");
  if (!timeDim) throw new Error("TIME_PERIOD dimension not found");

  // There should be exactly one series key when fetching a single tenor
  const seriesKey = Object.keys(series)[0];
  if (!seriesKey) return [];
  const observations = series[seriesKey].observations;

  return Object.entries(observations)
    .map(([idx, vals]) => {
      // Norges Bank returns values as strings inside the array e.g. ['4.172']
      const raw = (vals as (string | number | null)[])[0];
      const numVal = raw !== null && raw !== undefined ? parseFloat(String(raw)) : NaN;
      return {
        date: timeDim.values[parseInt(idx)].id,
        value: numVal,
      };
    })
    .filter((p) => isFinite(p.value))   // filters NaN and Infinity (covers null→NaN)
    .sort((a, b) => a.date.localeCompare(b.date));
}

// Internal maturity → dataset/instrument/tenorCode mapping
type MaturityCode = "3M" | "6M" | "1Y" | "2Y" | "5Y" | "10Y";

interface MaturitySpec {
  dataset: "GOVT_GENERIC_RATES" | "GOVT_ZEROCOUPON";
  tenorCode: string;       // as used in the API key
  instrument?: string;     // only for GOVT_GENERIC_RATES
}

const MATURITY_SPECS: Record<MaturityCode, MaturitySpec> = {
  "3M":  { dataset: "GOVT_GENERIC_RATES", tenorCode: "3M",  instrument: "TBIL" },
  "6M":  { dataset: "GOVT_GENERIC_RATES", tenorCode: "6M",  instrument: "TBIL" },
  "1Y":  { dataset: "GOVT_GENERIC_RATES", tenorCode: "12M", instrument: "TBIL" },
  "2Y":  { dataset: "GOVT_ZEROCOUPON",    tenorCode: "2Y" },
  "5Y":  { dataset: "GOVT_GENERIC_RATES", tenorCode: "5Y",  instrument: "GBON" },
  "10Y": { dataset: "GOVT_GENERIC_RATES", tenorCode: "10Y", instrument: "GBON" },
};

function buildUrl(spec: MaturitySpec, startDate?: string): string {
  const base = NB_BASE;
  let key: string;
  if (spec.dataset === "GOVT_GENERIC_RATES") {
    key = `B.${spec.tenorCode}.${spec.instrument}`;
  } else {
    // GOVT_ZEROCOUPON key format: B.{tenor}
    key = `B.${spec.tenorCode}`;
  }
  const params = new URLSearchParams({ "format": "sdmx-json", "locale": "en" });
  if (startDate) params.set("startPeriod", startDate);
  return `${base}/${spec.dataset}/${key}?${params}`;
}

export async function nbFetchYield(maturity: MaturityCode, startDate?: string): Promise<NBPoint[]> {
  const spec = MATURITY_SPECS[maturity];
  if (!spec) throw new Error(`Unknown maturity: ${maturity}`);
  return fetchSDMX(buildUrl(spec, startDate));
}

// Fetch multiple maturities and return flat array of {date, maturity, value}
export async function nbFetchAll(
  maturities: MaturityCode[],
  startDate?: string
): Promise<{ date: string; maturity: string; value: number }[]> {
  const results = await Promise.all(
    maturities.map((m) =>
      nbFetchYield(m, startDate)
        .then((pts) => pts.map((p) => ({ ...p, maturity: m })))
        .catch(() => [] as { date: string; maturity: string; value: number }[])
    )
  );
  return results.flat();
}

export const NB_CURVE_MATURITIES: MaturityCode[] = ["3M", "6M", "1Y", "2Y", "5Y", "10Y"];

File: backend/lib/norgesbank.py
"""Norges Bank SDMX-JSON API client — mirrors norgesbank.ts.

Datasets:
  GOVT_GENERIC_RATES  — TBIL (3M/6M/12M) and GBON (5Y/10Y), business daily
  GOVT_ZEROCOUPON     — zero-coupon yields (2Y), daily
"""
from typing import Optional

import httpx

NB_BASE = "https://data.norges-bank.no/api/data"

_MATURITY_SPECS: dict[str, dict] = {
    "3M":  {"dataset": "GOVT_GENERIC_RATES", "tenor": "3M",  "instrument": "TBIL"},
    "6M":  {"dataset": "GOVT_GENERIC_RATES", "tenor": "6M",  "instrument": "TBIL"},
    "1Y":  {"dataset": "GOVT_GENERIC_RATES", "tenor": "12M", "instrument": "TBIL"},
    "2Y":  {"dataset": "GOVT_ZEROCOUPON",    "tenor": "2Y"},
    "5Y":  {"dataset": "GOVT_GENERIC_RATES", "tenor": "5Y",  "instrument": "GBON"},
    "10Y": {"dataset": "GOVT_GENERIC_RATES", "tenor": "10Y", "instrument": "GBON"},
}


def _build_url(spec: dict, start_date: Optional[str] = None) -> str:
    if spec["dataset"] == "GOVT_GENERIC_RATES":
        key = f"B.{spec['tenor']}.{spec['instrument']}"
    else:
        key = f"B.{spec['tenor']}"
    params = "format=sdmx-json&locale=en"
    if start_date:
        params += f"&startPeriod={start_date}"
    return f"{NB_BASE}/{spec['dataset']}/{key}?{params}"


async def _fetch_sdmx(url: str) -> list[dict]:
    """Parse a Norges Bank SDMX-JSON response into [{date, value}] points."""
    async with httpx.AsyncClient(timeout=30) as client:
        r = await client.get(url)
        r.raise_for_status()

    json = r.json()
    struct = json["data"]["structure"]
    ds = json["data"]["dataSets"][0]
    series: dict = ds["series"]

    # Find the TIME_PERIOD observation dimension
    obs_dims: list[dict] = struct["dimensions"]["observation"]
    time_dim = next((d for d in obs_dims if d["id"] == "TIME_PERIOD"), None)
    if time_dim is None:
        raise ValueError("TIME_PERIOD dimension not found in Norges Bank response")

    # There should be exactly one series key when fetching a single tenor
    series_key = next(iter(series), None)
    if series_key is None:
        return []

    observations: dict = series[series_key]["observations"]
    points = []
    for idx_str, vals in observations.items():
        # Norges Bank returns values as strings: ['4.172'] — must parseFloat
        raw = vals[0] if vals else None
        if raw is None:
            continue
        try:
            num = float(str(raw))
        except (ValueError, TypeError):
            continue
        if not (num == num) or num in (float("inf"), float("-inf")):  # NaN / Inf check
            continue
        date_str = time_dim["values"][int(idx_str)]["id"]
        points.append({"date": date_str, "value": num})

    return sorted(points, key=lambda p: p["date"])


async def nb_fetch_yield(maturity: str, start_date: Optional[str] = None) -> list[dict]:
    spec = _MATURITY_SPECS.get(maturity)
    if spec is None:
        raise ValueError(f"Unknown maturity: {maturity}")
    return await _fetch_sdmx(_build_url(spec, start_date))


async def nb_fetch_all(maturities: list[str], start_date: Optional[str] = None) -> list[dict]:
    """Fetch multiple maturities; returns flat list of {date, maturity, value}."""
    import asyncio

    async def fetch_one(m: str) -> list[dict]:
        try:
            pts = await nb_fetch_yield(m, start_date)
            return [{"date": p["date"], "maturity": m, "value": p["value"]} for p in pts]
        except Exception:
            return []

    results = await asyncio.gather(*[fetch_one(m) for m in maturities])
    return [item for sublist in results for item in sublist]
